Convert  getinstancevariable to new backend IR

k0kubun opened the pull request:
I merged Maxime's register allocator fix  to @noahgibbs's branch and addressed another panic in `Opnd::mem`.

noahgibbs committed:
Convert getinstancevariable to new backend IR

k0kubun committed:
Merge remote-tracking branch 'origin/yjit_backend_ir' into noah_getinstancevariable

k0kubun committed:
Support mem-based mem

maximecb commented on yjit/src/backend/ir.rs:
```diff
@@ -242,6 +242,15 @@ impl Opnd
+                    disp: disp + base.disp,
+                    num_bits: num_bits,
+                })
+            },
```
Using a memory operand as the base for another memory operand is intentionally not supported

k0kubun replied:
Removed this in https://github.com/Shopify/ruby/pull/352/commits/86eb39dd784d6b39d6a01ccfd4ae4803e97d717c.

maximecb commented on yjit/src/backend/tests.rs:
```diff
@@ -112,6 +112,12 @@ fn test_mov_mem2mem()
+fn test_mem_based_mem() {
+    let self_asm_opnd = Opnd::mem(64, CFP, RUBY_OFFSET_CFP_SELF);
+    Opnd::mem(64, self_asm_opnd, RUBY_OFFSET_RBASIC_KLASS);
+}
```
Would rather not have to support this, because then we can have a long chain of memory operands that depend on each other.

k0kubun replied:
Ok, so we're supposed to use `asm.load` for this case, is that correct?

maximecb replied:
Yes

k0kubun replied:
Thanks, fixed as such in https://github.com/Shopify/ruby/pull/352/commits/86eb39dd784d6b39d6a01ccfd4ae4803e97d717c.

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -227,11 +227,14 @@ macro_rules! counted_exit {
 
             // Jump to the existing side exit
-            jmp_ptr(ocb, $existing_side_exit);
+            ocb_asm.jmp(Target::CodePtr($existing_side_exit));
```
Here you should be able to do `$existing_side_exit.into()`

k0kubun replied:
addressed in https://github.com/Shopify/ruby/pull/352/commits/93f02842ea74691481607812e21c5da1fdcdf516

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -1771,40 +1774,39 @@ fn gen_checkkeyword(
     match shape {
         BranchShape::Next0 | BranchShape::Next1 => unreachable!(),
-        BranchShape::Default => jnz_ptr(cb, target0),
+        BranchShape::Default => asm.jnz(Target::CodePtr(target0)),
```
`target0.into()`

k0kubun replied:
addressed in https://github.com/Shopify/ruby/pull/352/commits/93f02842ea74691481607812e21c5da1fdcdf516

maximecb commented on yjit/src/codegen.rs:
```diff
@@ -1927,63 +1937,43 @@ fn gen_get_ivar(
-        jle_ptr(cb, counted_exit!(ocb, side_exit, getivar_idx_out_of_range));
+        let num_slots = Opnd::mem(32, recv, ROBJECT_OFFSET_NUMIV);
+        asm.cmp(num_slots, Opnd::UImm(ivar_index as u64));
+        asm.jbe(Target::CodePtr(counted_exit!(ocb, side_exit, getivar_idx_out_of_range)));
```
We should probably have counted_exit directly return a `Target::CodePtr`

noahgibbs replied:
I tried that, but thought I should convert it all to the new Assembler interface or none of it. It just seems odd to return an Assembler-wrapped CodePtr, but to take and return raw CodePtrs for side_exit and use a CodeBlock not an Assembler.

k0kubun replied:
For now I just added https://github.com/Shopify/ruby/pull/352/commits/7a93d26ad7f46189c642da12e12778a63c2cd389

maximecb reviewed (commented):
Added a few comments. I think we should probably add a few getivar tests to `test_yjit_new_backend.rb` as well.

k0kubun committed:
Just load obj_opnd to a register

k0kubun committed:
Use more into()

k0kubun committed:
Add tests for getivar

k0kubun committed:
Flip the nil-out condition

k0kubun committed:
Apply another into()

k0kubun commented:
Addressed the comments, and fixed one thing that was probably a mistake https://github.com/Shopify/ruby/pull/352/commits/c3b78287880dc18bd480beb83ed1d0dcec73b25e. 

`CSelE` generates `cmovne`, and `CSelNE` generates `cmove`; the opposite of what IR says. Since we want to move Qnil when it's equal to Qundef, I believe it needs to be `cmove` (`CSelNE`, given how the IR is implemented).

noahgibbs commented:
> `CSelE` generates `cmovne`, and `CSelNE` generates `cmove`; the opposite of what IR says. Since we want to move Qnil when it's equal to Qundef, I believe it needs to be `cmove` (`CSelNE`, given how the IR is implemented).

I believe this is intentional. I had read CSelE as "Select if Equal", so you'd get the first item if the equal/zero flag is true, and the second one if not. Which turns into a MOV &lt;firstitem&gt; / CMOVNE &lt;seconditem&gt; on x86 to get the same effect.

noahgibbs commented on yjit/src/codegen.rs:
```diff
@@ -1993,47 +1983,45 @@ fn gen_get_ivar(
+        asm.comment("guard extended getivar");
+        let flags_opnd = Opnd::mem(64, recv, RUBY_OFFSET_RBASIC_FLAGS);
+        asm.test(flags_opnd, Opnd::UImm(ROBJECT_EMBED as u64));
         let side_exit = counted_exit!(ocb, side_exit, getivar_megamorphic);
```
I believe setting side_exit = counted_exit here and then using it below for another counted_exit can sometimes result in multiple stats being counted for the same side exit. In this case, I believe that's not intentional.

If we sub "counted_exit!(ocb, side_exit, getivar_megamorphic)" for side_exit in the call to jit_chain_guard but do *not* reassign side_exit, I believe that fixes the problem.

(You didn't introduce this bug, Kokubun, but I spotted it while reviewing.)

k0kubun replied:
like this https://github.com/Shopify/ruby/pull/352/commits/c725b865c89bbb50a1d2edfd422802fc890495db? I wish we didn't need to substitute anything, but it seems needed to avoid double mutable borrows in the same call.

noahgibbs replied:
Yes, like that. And yeah, I think that's how the bug got introduced originally - you can't just put the counted_exit! call inline in the function call because of the mutable borrows :-/

k0kubun committed:
Fix duplicated counts of side_exit

k0kubun commented:
I think I addressed all the past reviews, including two ivar test paths in test_yjit_new_backend.rb. Since I ended up relying on this branch from https://github.com/Shopify/ruby/pull/357 and some conflict resolution effort would be needed after squash merge, let me merge this one :bow: YJIT CIs are already passing, and I can address any further feedback after merging it. Thank you for having a look.
